libs/server/src/lib/api/executable.ts

```diff
@@ -0,0 +1,160 @@
+import { normalizeCommands } from '../utils/architect';
+import { FileUtils } from '../utils/file-utils';
+import { readJsonFile } from '../utils/utils';
+import { Commands } from './commands';
+import { createDetailedStatusCalculator } from './detailed-status-calculator';
+import { Telemetry } from '../telemetry';
+
+// todo(matt) should be handled by Commands
+let commandRunIndex = 0;
+
+export interface RunResult {
+  id: string;
+}
+
+export enum ExecutableType {
+  add = 'add',
+  new = 'new',
+  generate = 'generate',
+  npm = 'npm',
+  ng = 'ng'
+}
+
+// TODO(matt): next step of this refactoring is 1 level of inheritance to
+// handle differences in name / path behaviour for different executables
+export class Executable {
+  private readonly isNvm: boolean;
+
+  private _path = '';
+  private _name: string;
+
+  get name() {
+    return this.isNvm ? 'nvm' : this._name;
+  }
+
+  get path() {
+    return this.isNvm ? `nvm exec ${this._path}` : this._path;
+  }
+
+  setPath(path: string) {
+    this._path = path;
+  }
+
+  setName(name: string) {
+    this._name = name;
+  }
+
+  hasPath(): boolean {
+    return this._path !== '';
+  }
+
+  constructor(
+    name: string,
+    private readonly telemetry: Telemetry,
+    private readonly buildTerminal: PseudoTerminalFactory,
+    private readonly fileUtils: FileUtils,
+    private readonly commands: Commands
+  ) {
+    this._name = name;
+    this.isNvm = fileUtils.supportsNvm() && fileUtils.useNvm();
+  }
+
+  run(
+    type: ExecutableType,
+    cwd: string,
+    cmds: string[],
+    addToRecent: boolean = true
+  ): RunResult {
+    if (!this.hasPath()) {
+      throw new Error('setPath of this executable before running');
+    }
+
+    const workspace = workspaceName(type, cwd);
+    const statusCalculator = createDetailedStatusCalculator(cwd, cmds);
+
+    const id = `${this.name} ${cmds.join(' ')} ${commandRunIndex++}`;
+    let command = `${this.path} ${cmds.join(' ')}`;
+
+    // We currently don't support the windows implementation of NVM.
+    if (this.isNvm) {
```

Comment:
You aren't doing a windows check here so the comment above doesn't make sense

Reply:
Yeah, I preserved that from the existing code, I think the intention is to explain why that check isn't there. Can remove if you like

Reply:
Meh, keep it